### PR TITLE
Add assert_eq! to special-cased macros

### PR DESCRIPTION
**Pull request by dtwood**
Continuing with the implementation of rust-lang-nursery/fmt-rfcs#86, allows for this form of `assert_eq!` macros:
```rust
assert_eq!(
    left.id, right.id,
    "IDs are not equal: {:?} {:?}",
    left, right
);
```

As a side-effect of how I implemented this it also allows for `assert!` macros to have the format arguments split across multiple lines even if the assert condition is not simple:
```rust
assert!(
    result >= 42,
    "The result must be at least 42: {:?}",
    result, result.code, context
);
```
I think that's reasonable behaviour, although not explicitly specified in the RFC pull request (rust-lang-nursery/fmt-rfcs#110)

### CODE REVIEW

**Comment by dtwood** [comment]
I'm a bit concerned by my new return code from `maybe_get_args_offset` (`Option<(bool, bool, usize)>`), having a reasonably large number of items inside the tuple (the same with `DefinitiveListTactic::SpecialMacro`), but hopefully it's relatively clear what they each mean.

File: src/lists.rs [hunk 1]
```diff
@@ -313,26 +311,33 @@ where
...
                     // Nothing
-                } else if second {
+                } else if i < num_args_before {
+                    if one_line_before {
+                        result.push(' ');
+                    } else {
+                        result.push('\n');
+                        result.push_str(indent_str);
+                    }
+                } else if i == num_args_before {
```
Comment (topecongiro) [inline src/lists.rs]: Would you please combine `i == num_args_before` and `i == num_args_before + 1` into a single `else if` block?

File: src/expr.rs [hunk 2]
```diff
@@ -1811,25 +1811,34 @@ fn rewrite_string_lit(context: &RewriteContext, span: Span, shape: Shape) -> Opt
     )
 }
 
-const FORMAT_LIKE_WHITELIST: &[&str] = &[
+const SPECIAL_MACRO_WHITELIST: &[(&str, usize)] = &[
```
Comment (topecongiro) [inline src/expr.rs]: Would you please add a comment on what `usize` represents?

**Comment by dtwood** [comment]
Thanks for the review 🙂, I've made those two changes

**Comment by nrc** [comment]
> As a side-effect of how I implemented this it also allows for assert! macros to have the format arguments split across multiple lines even if the assert condition is not simple:

I think that this is specified, though perhaps not clearly enough. If the arguments are not simple, then they should be on one line each.

I.e., there are three ways the assert_eq macro can be laid out:

```
// single-line:
assert_eq!(a, b, "foo {} {}", c, d);
// special-case
assert_eq!(
    a, b,
    "foo {} {}",
    c, d,
);
// multi-line
assert_eq!(
    a,
    b,
    "foo {} {}",
    c,
    d,
);
```

The single-line format can be used if the macro fits inside max-width and the function args heuristic width. If not, then the special-case can be used if `c` and `d` are simple. If they are not simple and we can't use the single-line format, then we must use the multi-line format.

**Comment by dtwood** [comment]
I've updated to only have those three formats

**Comment by nrc** [comment]
Thank you!

**Pull request merged**